module/VuFindTheme/src/VuFindTheme/View/Helper/ConcatTrait.php

```diff
@@ -228,15 +231,62 @@ protected function getConcatenatedFilePath($group)
         $filename = md5($group['key']) . '.min.' . $this->getFileType();
         $concatPath = $this->getResourceCacheDir() . $filename;
         if (!file_exists($concatPath)) {
-            $minifier = $this->getMinifier();
-            foreach ($group['items'] as $item) {
-                $details = $this->themeInfo->findContainingTheme(
-                    $this->getFileType() . '/' . $this->getResourceFilePath($item),
-                    ThemeInfo::RETURN_ALL_DETAILS
-                );
-                $minifier->add($details['path']);
+            $pid = getmypid();
+            $lockfile = "$concatPath.lock";
+            $handle = fopen($lockfile, 'c+');
+            if (!is_resource($handle)) {
+                throw new \Exception("Could not open lock file $lockfile");
+            }
+            if (!flock($handle, LOCK_EX)) {
+                fclose($handle);
+                throw new \Exception("Could not lock file $lockfile");
+            }
+            // Check again if file exists after acquiring the lock
+            if (!file_exists($concatPath)) {
+                $startTime = microtime(true);
+                try {
+                    $tempfile = "$concatPath.tmp";
+                    foreach ($group['items'] as $item) {
+                        $details = $this->themeInfo->findContainingTheme(
+                            $this->getFileType() . '/'
+                            . $this->getResourceFilePath($item),
+                            ThemeInfo::RETURN_ALL_DETAILS
+                        );
+                        if ($this->isMinifiable($details['path'])) {
```

Comment:
Would it make sense to refactor lines 255-273 into a separate getMinifiedData method? That way, you could extend the method in the HeadScript helper to put the JS-specific logic there, instead of having a special case in the shared logic. It might also help readability a bit, as this method has become rather long.